week-06/assessment/field-agent/src/main/java/learn/field_agent/domain/SecurityClearanceService.java

```diff
@@ -1,2 +1,100 @@
-package learn.field_agent.domain;public class SecurityClearanceService {
+package learn.field_agent.domain;
+
+import learn.field_agent.data.SecurityClearanceRepository;
+import learn.field_agent.models.SecurityClearance;
+import org.springframework.stereotype.Service;
+
+import java.util.List;
+
+@Service
+public class SecurityClearanceService {
+    private final SecurityClearanceRepository repository;
+
+    public SecurityClearanceService(SecurityClearanceRepository repository) {
+        this.repository = repository;
+    }
+
+    public SecurityClearance findById(int securityClearanceId) {
+
+        return repository.findById(securityClearanceId);
+    }
+
+    public List<SecurityClearance> findAll() {
+        return repository.findAll();
+    }
+
+    public Result<SecurityClearance> add(SecurityClearance securityClearance) {
+        Result<SecurityClearance> result = validate(securityClearance);
+        if (!result.isSuccess()) {
+            return result;
+        }
+
+        // Check for duplicate name, needed findByName() method to validate.
+        if (repository.findByName(securityClearance.getName()) != null) {
+            result.addMessage("This security clearance exists.", ResultType.INVALID);
+            return result;
+        }
+
+        if (securityClearance.getSecurityClearanceId() != 0) {
+            result.addMessage("Cannot set security clearance ID for 'add' operation", ResultType.INVALID);
+            return result;
+        }
+
+        SecurityClearance addedSecurityClearance = repository.add(securityClearance);
+        if (addedSecurityClearance != null) {
+            result.setPayload(addedSecurityClearance);
+        } else {
+            result.addMessage("Failed to add Security Clearance.", ResultType.INVALID);
+        }
+
+        return result;
+    }
+    public Result<SecurityClearance> update(SecurityClearance securityClearance) {
+        Result<SecurityClearance> result = validate(securityClearance);
+        if (!result.isSuccess()) {
+            return result;
+        }
+        SecurityClearance existingSecurityClearance = repository.findById(securityClearance.getSecurityClearanceId());
+
+        // Check if the name is being changed to the same name
+        if (existingSecurityClearance.getName().equals(securityClearance.getName())) {
```

Comment:
Ooo. Methodical consideration of:
1. Checking that the name has or hasn't changed.
2. If changed, check for a duplicate name.
💯